remill/Arch/X86/Semantics/UNCOND_BR.cpp

```diff
@@ -22,6 +22,28 @@ DEF_SEM(JMP, T target_pc) {
   return memory;
 }
 
+template <typename T>
+DEF_SEM(JMP_FAR_MEM, T target_seg_pc) {
+  HYPER_CALL = AsyncHyperCall::kX86JmpFar;
+  uint64_t target_fword = UShr(UShl(Read(target_seg_pc), 0xf), 0xf);
+  auto pc = static_cast<uint32_t>(target_fword);
+  auto seg = static_cast<uint16_t>(UShr(target_fword, 32));
+  WriteZExt(REG_PC, pc);
+  Write(REG_CS.flat, seg);
+  return __remill_async_hyper_call(state, REG_PC, memory);
```

Comment:
Here you want is to just `return memory`. Then use of a `__remill_async_hyper_call` should be done by a higher level, i.e. the code doing the lifting. You'll need to modify the `CreateCategory`, and possibly `IsIndirectJump` and `IsIndirectCall`, maybe adding some new function that checks for the far calls/jumps only, and returns `kCategoryAsyncHyperCall`.

The idea here is that, it's hard to make a function call inside of a semantics function into a terminating tail call after all optimizations are said and done. Semantics should be mostly come in and come out, with the actual redirecting of control flow happing between/outside of them. At that level, you can say "this is an async hyper call, so call `intrinsics.async_hyper_call`".

Reply:
That makes sense to me, PR updated.